libursa/src/signatures/bls.rs

```diff
@@ -0,0 +1,693 @@
+        // `Signature` should be generated by calling `Signature::new_with_rk_mitigation`
+        // to avoid rogue key attacks. If using proof of possession mitigation
+        // then `Signature` can be generated by calling `Signature::new`
+        pub fn new(signatures: &[Signature]) -> Self {
+            AggregatedSignature(signatures.iter().fold(G2::identity(), |sig, s| sig + &s.0))
+        }
+
+        // Verify with rogue key attack mitigation.
+        pub fn verify(&self, message: &[u8], pk: &AggregatedPublicKey) -> bool {
+            let point = G2::from_msg_hash(message);
+            GT::ate_2_pairing(&-&G1::generator(), &self.0, &pk.0, &point).is_one()
+        }
+
+        // Verify without rogue key mitigation. Assumes caller has handled
+        // rogue key mitigation some other way like proof of possession.
+        // This practice is discouraged in favor of the other method
+        // but there are use cases where proof of possession is better suited
+        pub fn verify_no_rk(&self, message: &[u8], pks: &[PublicKey]) -> bool {
+            let apk = pks.iter().fold(G1::identity(), |a, p| a + &p.0);
+            let point = G2::from_msg_hash(message);
+            GT::ate_2_pairing(&-&G1::generator(), &self.0, &apk, &point).is_one()
+        }
+
+        /// This should be used to verify quickly multiple BLS aggregated signatures by batching
+    /// versus verifying them one by one as it reduces the number of computed pairings
+        pub fn batch_verify(inputs: &[(&[u8] /* message */, &AggregatedSignature, &AggregatedPublicKey)]) -> bool {
+            // To combat the rogue key attack and avoid checking for distinct messages
+            // use batch verification as described in the end of section 3.1 from https://eprint.iacr.org/2018/483
+            let mut pairs = Vec::new();
+            let mut sig = G2::identity();
+            for (msg, asg, apk) in inputs {
+                let random_exponent = FieldElement::random();
+                let g2 = G2::from_msg_hash(msg);
+                sig += &asg.0 * &random_exponent;
+                pairs.push((&apk.0 * &random_exponent, g2));
+            }
+
+            pairs.push((-&G1::generator(), sig));
+
+            let ate_pairs = pairs.iter().map(|(g1, g2)| (g1, g2)).collect();
+            GT::ate_multi_pairing(ate_pairs).is_one()
+        }
+    }
+
+    /// Creates a new BLS key pair
+    pub fn generate() -> (PublicKey, PrivateKey) {
+        let sk = PrivateKey::random();
+        let pk = PublicKey::new(&sk);
+        (pk, sk)
+    }
+
+    bls_tests_impl!();
+}
+
+/// This version is the small BLS signature scheme
+/// with the public key group in G2 and signature group in G1.
+/// 97 bytes signatures and 192 byte public keys
+///
+/// This results in smaller signatures but slower operations and bigger public key.
+/// This is good for situations where space is a consideration and verification is infrequent
+pub mod small {
```

Comment:
Rather than duplicating the same logic as present in `mod normal` in here, you can use macros if you want to use both configurations in the same compiled code. Or use features that put signature in G1 or G2 if you want the configuration to be picked at compile time

Reply:
Except that the math is slightly different in the verify steps. I've tried to make many things common with macros and ultimately didn't work out well and made the code more confusing.

Reply:
The math should not be different. Can you point out the specific instance?

Reply:
https://github.com/mikelodder7/ursa/blob/bls/libursa/src/signatures/bls.rs#L372
https://github.com/mikelodder7/ursa/blob/bls/libursa/src/signatures/bls.rs#L618

They have to be different because the pairing expects the values to be G1, G2. Just using a macro to change G1->G2 and G2->G1 doesn't work there because you have to swap the variables too.

Reply:
The math isn't necessarily different. Its just an ordering of the variables.

Reply:
The macro takes a function as one of the arguments which does the pairing. One function's signature will be something like `Fn(&G1, &G2, &G1, &G2)` and the other's will be `Fn(&G2, &G1, &G2, &G1)` but both call the pairing function. Or you can keep the signature same but change the order of params in the implementation.